### PR TITLE
added 12th HW

### PR DESCRIPTION
**Pull request by 415816**

### CODE REVIEW

File: lec_12_HW/task_1.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// 1. Создайте дженерик функцию getFirstElement, которая принимает массив элементов типа T, и возвращает первый элемент (типа T).
+
+function getFirstElement<T>(array: T[]) {
+    return array[1];
```
Comment (VladimirLQA) [inline lec_12_HW/task_1.ts]: `JavaScript arrays are zero-indexed: the first element of an array is at index 0, the second is at index 1, and so on — and the last element is at the value of the array's length property minus 1.`

File: lec_12_HW/task_2.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// 1. Напишите функцию, реализующую методы массивов map. Функции принимают на вход массив и колбэк. Используйте дженерик типы.
+//    Затипизировать надо саму функцию и коллбэк.
+//    Создать реализацию функции map, принимающую массив чисел 1-5, возвращающую новый массив,
+//    где каждый каждый элемент - это элемент исходного массива умноженный на его индекс
+//    Пример:
+//    map([1,2,3,4,5], callback) => [0,2,6,12,20]
+
+function map<T, P>(array: T[], callback: (elem: T, index: number) => P): P[] {
+    let result: P[] = [];
```
Comment (VladimirLQA) [inline lec_12_HW/task_2.ts]:
```suggestion
    const result: P[] = [];
```

File: lec_12_HW/task_2.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+// 1. Напишите функцию, реализующую методы массивов map. Функции принимают на вход массив и колбэк. Используйте дженерик типы.
+//    Затипизировать надо саму функцию и коллбэк.
+//    Создать реализацию функции map, принимающую массив чисел 1-5, возвращающую новый массив,
+//    где каждый каждый элемент - это элемент исходного массива умноженный на его индекс
+//    Пример:
+//    map([1,2,3,4,5], callback) => [0,2,6,12,20]
+
+function map<T, P>(array: T[], callback: (elem: T, index: number) => P): P[] {
```
Comment (VladimirLQA) [inline lec_12_HW/task_2.ts]:
```suggestion
function map<T, P>(array: T[], callback: (elem: T, index?: number, oArray?: readonly T[]) => P): P[] {
```

File: lec_12_HW/task_2.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+//   В случае если ключи повторяются, значение в объекте должно быть последним из встречающихся.
+
+//   Требования:
+//     - Функция должна быть дженерик и работать с любыми типами значений.
+//     - Функция должна корректно обрабатывать массив пар, включая случаи, когда ключи повторяются.
+type Pair<T> = [string, T];
+type ObjectResult<T> = { [key: string]: T };
+
+function generateObject<T>(array: Pair<T>[]): ObjectResult<T> {
+	let result: ObjectResult<T> = {};
```
Comment (VladimirLQA) [inline lec_12_HW/task_2.ts]:
```suggestion
	const result: ObjectResult<T> = {};
```

File: lec_12_HW/task_2.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+	for (let i = 0; i < array.length; i++) {
```
Comment (VladimirLQA) [inline lec_12_HW/task_2.ts]: As alternative
```js
 for (const [key, value] of array) {
   result[key] = value;
}
return result;
// or

return Object.fromEntries(array)
```

**Review by VladimirLQA — APPROVED** [review]

File: lec_12_HW/task_1.ts (hunk 1 above)
Comment (415816) [reply lec_12_HW/task_1.ts]: Ok...

File: lec_12_HW/task_2.ts (hunk 2 above)
Comment (415816) [reply lec_12_HW/task_2.ts]: Ok.

File: lec_12_HW/task_2.ts (hunk 3 above)
Comment (415816) [reply lec_12_HW/task_2.ts]: Ok.

File: lec_12_HW/task_2.ts (hunk 4 above)
Comment (415816) [reply lec_12_HW/task_2.ts]: Ok.

File: lec_12_HW/task_2.ts (hunk 5 above)
Comment (415816) [reply lec_12_HW/task_2.ts]: Oh, cool, thank you

**Pull request merged**